### PR TITLE
BRAYNS 314 - Flatmap areas

### PR DESCRIPTION
**Pull request by NadirRoGue**

### CODE REVIEW

File: plugins/AtlasExplorer/api/IUseCase.h [hunk 1]
```diff
@@ -23,17 +23,31 @@
...
+     */
     virtual std::string getName() const = 0;
 
-    virtual bool isVolumeValid(const AtlasVolume &volume) const = 0;
+    /**
+     * @brief Tests wether the use case can be applied to the atlas.
+     * @param atlas Atlas to be tested.
+     * @return true If the atlas is valid for this use case, false otherwise.
+     */
+    virtual bool isAtlasValid(const Atlas &atlas) const = 0;
```
Comment (Adrien4193) [inline plugins/AtlasExplorer/api/IUseCase.h:44]: isValidAtlas ?

File: plugins/AtlasExplorer/api/usecases/FlatmapAreas.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+{
+    return atlas.getVoxelType() == VoxelType::flatmap;
+}
+
+std::unique_ptr<brayns::Model> FlatmapAreas::run(const Atlas &atlas, const brayns::JsonValue &payload) const
+{
+    (void)payload;
+    assert(dynamic_cast<const FlatmapAtlas *>(&atlas));
+
+    auto &flatmapAtlas = static_cast<const FlatmapAtlas &>(atlas);
```
Comment (Adrien4193) [inline plugins/AtlasExplorer/api/usecases/FlatmapAreas.cpp:171]: Maybe at some point we can make a utility function like cast<T>(value) {assert(dynamic), return static}.

File: plugins/AtlasExplorer/api/usecases/OutlineShell.cpp [hunk 3]
```diff
@@ -20,100 +20,47 @@
...
-        for (size_t i = 0; i < data.size(); ++i)
+        for (size_t i = 0; i < voxelCount; ++i)
         {
-            auto isEmpty = data[i] == minValue;
-
-            if (!std::isfinite(data[i]) || isEmpty)
+            if (atlas.isValidVoxel(i))
             {
-                continue;
+                result[i] = 255u;
```
Comment (Adrien4193) [inline plugins/AtlasExplorer/api/usecases/OutlineShell.cpp:46]: 255u is uint32_t no ? Could we use std::numeric_limits<uint8> or ~0 ?

File: plugins/AtlasExplorer/io/nrrdloader/data/decoders/AsciiDecoder.cpp [hunk 4]
```diff
@@ -34,13 +33,16 @@ class AsciiParser
...
     {
-        std::vector<T> result(tokens.size());
-
-        for (size_t i = 0; i < tokens.size(); ++i)
+        std::vector<T> result;
+        result.reserve(tokenCount);
+        while (!input.empty())
         {
-            brayns::StringParser<T>::parse(tokens[i], result[i]);
+            auto token = brayns::StringExtractor::extractUntilOneOf(input, brayns::StringConstants::tokenDelimiters);
```
Comment (Adrien4193) [inline plugins/AtlasExplorer/io/nrrdloader/data/decoders/AsciiDecoder.cpp:42]: extractToken ?

File: plugins/AtlasExplorer/io/nrrdloader/data/decoders/HexDecoder.cpp [hunk 5]
```diff
@@ -31,22 +33,19 @@ class HexContentParser
...
         std::string result;
         result.reserve(expectedResultSize);
 
         std::string hexBuffer(3, '\0');
         size_t hexIndex = 0;
 
-        for (auto token : tokens)
+        while (!input.empty())
         {
+            auto token = brayns::StringExtractor::extractUntilOneOf(input, brayns::StringConstants::tokenDelimiters);
```
Comment (Adrien4193) [inline plugins/AtlasExplorer/io/nrrdloader/data/decoders/HexDecoder.cpp:47]: extractToken ?

**Review by Adrien4193 — APPROVED** [review]

File: plugins/AtlasExplorer/api/IUseCase.h (hunk 1 above)
Comment (NadirRoGue) [reply plugins/AtlasExplorer/api/IUseCase.h:44]: For me it goes on how would you ask the question:
Is the atlas valid?
or
Is valid the atlas?
To me it makes more sense isAtlasValid

File: plugins/AtlasExplorer/api/usecases/OutlineShell.cpp (hunk 3 above)
Comment (NadirRoGue) [reply plugins/AtlasExplorer/api/usecases/OutlineShell.cpp:46]: Will change for numeric limits

File: plugins/AtlasExplorer/io/nrrdloader/data/decoders/AsciiDecoder.cpp (hunk 4 above)
Comment (NadirRoGue) [reply plugins/AtlasExplorer/io/nrrdloader/data/decoders/AsciiDecoder.cpp:42]: extractToken extracts only space-separated tokens, whereas the nrrd files have more delimiters. And I dont think its correct to use the list of delimiters from a specific file format into a generic function that is opaque to the user, even more when there is API like extractUntilOneOf which does what its needed.

File: plugins/AtlasExplorer/io/nrrdloader/data/decoders/AsciiDecoder.cpp [hunk 6]
```diff
@@ -34,13 +33,16 @@ class AsciiParser
...
+            auto &value = result.emplace_back();
+            brayns::StringParser<T>::parse(token, value);
```
[the remark this answers is not in the document]
Comment (NadirRoGue) [reply plugins/AtlasExplorer/io/nrrdloader/data/decoders/AsciiDecoder.cpp:44]: I think that would mix classes responsabilities

File: plugins/AtlasExplorer/io/nrrdloader/data/decoders/HexDecoder.cpp [hunk 7]
```diff
@@ -31,22 +33,19 @@ class HexContentParser
...
     static std::string toBinary(std::string_view input)
     {
-        auto tokens = Tokenizer::fromView(input);
-
-        size_t expectedResultSize = 0;
-        for (auto token : tokens)
-        {
-            expectedResultSize += token.size();
-        }
+        auto tokenCount = brayns::StringCounter::countOneOf(input, brayns::StringConstants::tokenDelimiters);
```
[the remark this answers is not in the document]
Comment (NadirRoGue) [reply plugins/AtlasExplorer/io/nrrdloader/data/decoders/HexDecoder.cpp:36]: countTokens count space-separated tokens.

File: plugins/AtlasExplorer/io/nrrdloader/data/decoders/HexDecoder.cpp (hunk 5 above)
Comment (NadirRoGue) [reply plugins/AtlasExplorer/io/nrrdloader/data/decoders/HexDecoder.cpp:47]: extractToken extracts only space-separated tokens, whereas the nrrd files have more delimiters. And I dont think its correct to use the list of delimiters from a specific file format into a generic function that is opaque to the user, even more when there is API like extractUntilOneOf which does what its needed.

File: plugins/AtlasExplorer/network/entrypoints/VisualizeAtlasUseCaseEntrypoint.cpp [hunk 8]
```diff
@@ -41,14 +65,23 @@ std::string VisualizeAtlasUseCaseEntrypoint::getDescription() const
...
-    request.reply(*instance);
+    auto &instance = brayns::ExtractModel::fromId(_models, modelId);
+    auto &model = instance.getModel();
+    auto &component = ExtractAtlas::fromModel(model);
+    auto &atlas = *component.atlas;
+
+    auto &useCase = _useCases.getUseCase(useCaseName);
+    if (!useCase.isAtlasValid(atlas))
+    {
+        throw brayns::JsonRpcException("The use-case is not valid for the given type of atlas");
```
[the remark this answers is not in the document]
Comment (NadirRoGue) [reply plugins/AtlasExplorer/network/entrypoints/VisualizeAtlasUseCaseEntrypoint.cpp:80]: Will change

**Review by Adrien4193 — DISMISSED** [review]

**Review by Adrien4193 — APPROVED** [review]

**Pull request merged**